### PR TITLE
Oceanwater 525 refactor arm services to use moveit based on joint trajectory controller interface

### PR DESCRIPTION
**Pull request by Samahu**
## Linked Issues:
Linked PR: https://github.com/nasa/ow_autonomy/pull/12
Epic Link: https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-518
Current Story: https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-525
Related Issues:
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-530
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-536
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-537
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-483 (api enhacements: remove delete_prev_traj)
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-526 (remove dependence on trajectory recording)
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-524
* https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-541

## Summary of Changes
* Refactored some of the arm services to bypass the trajectory recording functionality and rely completely on joint trajectory controller interface. Successfully ported services so far are:
  - **Stow**
  - **Unstow**
  - **GuardedMove**
  - **DeliverSample**
* Additionally, the code of **Stow** and **Unstow** services has been simplified by making use of the predefined poses.
* Removed delete_prev_traj from these services definitions as part of [OCEANWATER-483](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-483)

>Note (IMPORTANT): This task was split into 2 issues (linked below), the reason is the previous implementation was relying on fake controllers for the planning and then uses a disjointed set of controllers to execute the motion. This is no longer the case with the current configuration. With the current configuration joints now are part of single move group and joints can't belong to two active controllers at the same time. One way to solve this is enable switching between controllers at runtime as described in [OCEANWATER-530](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-530).

## Test
Currently only the euroa_terminator works out of the box with the new configuration, the remaining worlds are being worked on as part of [OCEANWATER-524](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-524).  
To verify the functionality of this ticket launch the eurpoa_terminator:
>Note: Enable the _quick_ solver on europa_terminator world before running this. A [problem](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-541) was observed when employing the _world_ - which is the current one - with the current configuration. Dealing with this issue outside the scoop of this ticket.
```bash
roslaunch ow europa_terminator.launch
```
Then either rqt_service_caller interface or launch a separate terminal (with the workspace sourced) and invoked any of the ported commands
```bash
rosservice call /arm/stow
rosservice call /arm/unstow
rosservice call /arm/deliver_sample ...
rosservice call /arm/guarded_move ...
```

### CODE REVIEW

File: ow_lander/srv/GuardedMove.srv [hunk 1]
```diff
@@ -1,13 +1,11 @@
 bool use_defaults           # Set to true to use the default touch location
-bool delete_prev_traj       # Set to true to cleanup ~/.ros
```
Comment (damianacatanoso) [inline ow_lander/srv/GuardedMove.srv:2]: If you remove delete_prev_traj from the .srv file, you should remove it from activity_guarded_move.py>arg_parsing , otherwise it will look for the value of this variable, which doesn't exist, and the service doesn't run. Same is valid for deliver_sample.

File: ow_lander/scripts/path_planning_commander.py [hunk 2]
```diff
@@ -25,59 +25,81 @@
...
+    self.arm_move_group.go(stowed_goal, wait=True)
+    self.arm_move_group.stop()
+    return True, "Done"
+
+  # === SERVICE ACTIVITIES - Unstow =============================
+  def handle_unstow(self, req):
+    """
+    :type req: class 'ow_lander.srv._Unstow.UnstowRequest'
+    """
+    stowed_goal = self.arm_move_group.get_named_target_values("arm_unstowed")
```
Comment (damianacatanoso) [inline ow_lander/scripts/path_planning_commander.py:61]: "stowed_goal" should be renamed "unstowed_goal". This unstowed configuration is different from the previous one.

Comment (Samahu) [reply ow_lander/scripts/path_planning_commander.py:61]: It uses the proper configuration, although the name of the variable is misleading. I will address that soon.

**Comment by damianacatanoso** [comment]
> Note: Enable the _quick_ solver on europa_terminator world before running this. A [problem](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-541) was observed when employing the _world_ - which is the current one - with the current configuration. Dealing with this issue outside the scoop of this ticket.
> 
This is a physics instability problem. It might be solvable with parameters tuning. I have been studying this problem. I can take care of this ticket if you wish.

File: ow_lander/scripts/path_planning_commander.py (hunk 2 above)
Comment (damianacatanoso) [reply ow_lander/scripts/path_planning_commander.py:61]: Thanks. Still the arm configuration corresponding to "unstow" in this branch is different from the one in melodic-devel. Here the arm goes lower and the angles between links is not 90 degrees.

File: ow_lander/srv/GuardedMove.srv (hunk 1 above)
Comment (Samahu) [reply ow_lander/srv/GuardedMove.srv:2]: addressed

File: ow_lander/scripts/path_planning_commander.py [hunk 3]
```diff
@@ -25,59 +25,81 @@
...
+class PathPlanningCommander(object):
   
   def __init__(self):
-    super(MoveGroupPythonInteface, self).__init__()
+    super(PathPlanningCommander, self).__init__()
     moveit_commander.roscpp_initialize(sys.argv)
     robot = moveit_commander.RobotCommander()
     scene = moveit_commander.PlanningSceneInterface()
 
     move_arm = moveit_commander.MoveGroupCommander("arm")
```
Comment (damianacatanoso) [inline ow_lander/scripts/path_planning_commander.py:36]: Please make sure that this group always uses RRTstar optimized for MechanicalWork.

**Comment by damianacatanoso** [comment]
When running "stow" I get error:
[ WARN] [1605820189.718456935, 1916524839.670000076]: Controller arm_controller failed with error GOAL_TOLERANCE_VIOLATED: 
[ WARN] [1605820189.718527851, 1916524839.670000076]: Controller handle arm_controller reports status ABORTED

File: ow_lander/scripts/path_planning_commander.py [hunk 4]
```diff
@@ -25,59 +25,81 @@
...
-    move_limbs = moveit_commander.MoveGroupCommander("limbs")
-    move_grinder = moveit_commander.MoveGroupCommander("grinder")
```
Comment (damianacatanoso) [inline ow_lander/scripts/path_planning_commander.py:39]: Why do you eliminate these two groups?

File: ow_lander/scripts/path_planning_commander.py (hunk 2 above)
Comment (Samahu) [reply ow_lander/scripts/path_planning_commander.py]: joint values revised for arm_unstowed pose

**Comment by Samahu** [comment]
> When running "unstow" I get error:
> [ WARN] [1605820189.718456935, 1916524839.670000076]: Controller arm_controller failed with error GOAL_TOLERANCE_VIOLATED:
> [ WARN] [1605820189.718527851, 1916524839.670000076]: Controller handle arm_controller reports status ABORTED
> As a consequence, the gazebo arm doesn't move.

There is a ticket to address as part of the epic .. it is the one about the trajectory constraint values

File: ow_lander/scripts/path_planning_commander.py (hunk 4 above)
Comment (Samahu) [reply ow_lander/scripts/path_planning_commander.py:39]: These would need to be done as part of OCEANWATER-536 and OCEANWATER-537 after ticket OCEANWATER-530 has been implemented

File: ow_lander/config/lander.srdf [hunk 5]
```diff
@@ -47,11 +47,11 @@
...
     <group_state name="arm_unstowed" group="arm">
-        <joint name="j_dist_pitch" value="0.7171" />
-        <joint name="j_hand_yaw" value="0" />
+        <joint name="j_dist_pitch" value="0" />
+        <joint name="j_hand_yaw" value="1.5707" />
         <joint name="j_prox_pitch" value="-1.5707" />
         <joint name="j_scoop_yaw" value="0" />
-        <joint name="j_shou_pitch" value="0.8537" />
+        <joint name="j_shou_pitch" value="1.5707" />
         <joint name="j_shou_yaw" value="0" />
```
Comment (damianacatanoso) [inline ow_lander/config/lander.srdf:55]: Why did you change these values?

Comment (Samahu) [reply ow_lander/config/lander.srdf:55]: These would resolve the issue you raised earlier about the unstowned pose not being inline with the previous configuration

**Comment by Samahu** [comment]
Is this PR good to go? I got approvals only on **ow_autonomy** counterpart but not on **ow_simulator**

File: ow_lander/srv/GuardedMove.srv (hunk 1 above)
Comment (mogumbo) [reply ow_lander/srv/GuardedMove.srv:2]: $ rosservice call /arm/guarded_move "{use_defaults: true, x: 0.0, y: 0.0, z: 0.0, direction_x: 0.0, direction_y: 0.0,
  direction_z: 0.0, search_distance: 0.0}" 
ERROR: service [/arm/guarded_move] responded with an error: error processing request: global name 'delete_prev_traj' is not defined

Is this the same delete_prev_traj problem that @damianacatanoso had? I still have it and my branches are up-to-date. I also get the same error from deliver_sample with use_defaults = true.

Comment (Samahu) [reply ow_lander/srv/GuardedMove.srv:2]: I get the same error. Let me check what is the issue!

Comment (Samahu) [reply ow_lander/srv/GuardedMove.srv:2]: I probably didn't test this part of the code as I should've, so really thanks! I have pushed needed changes to address the problem!

**Comment by mogumbo** [comment]
If I start the sim and unstow, the arm moves to its unstowed position and then bounces (looks like physics instability). I cannot stow from this point, but I can do guarded_move or deliver_sample. After any other move I can stow. Subsequent unstow moves do not exhibit the bouncing. I also see occasional bouncing while the arm is moving on all of the trajectories.

guarded_move never finds the ground, but I see a lot of LCP errors from ODE after it touches the ground.

I don't know if any of this stuff should be fixed in this ticket or other tickets.

**Comment by Samahu** [comment]
> If I start the sim and unstow, the arm moves to its unstowed position and then bounces (looks like physics instability). I cannot stow from this point, but I can do guarded_move or deliver_sample. After any other move I can stow. Subsequent unstow moves do not exhibit the bouncing. I also see occasional bouncing while the arm is moving on all of the trajectories.
> 
> guarded_move never finds the ground, but I see a lot of LCP errors from ODE after it touches the ground.
> 
> I don't know if any of this stuff should be fixed in this ticket or other tickets.

I didn't observe this behavior or maybe didn't pay much attention to it. Just note that the PID values were not optimized yet.

With regard to guarded_move the not being detected is due to the fact that the script is not actually being executed at the moment. The ground_detection is within the trajectory_publisher.py which is being phased out. Of course I have a task to deal with this already: https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-540

**Comment by mogumbo** [comment]
> I didn't observe this behavior or maybe didn't pay much attention to it. Just note that the PID values were not optimized yet
Do you think the bouncing of the arm would cause stow to not execute? This seems like a separate problem that should be addressed to make the code more robust.

**Comment by Samahu** [comment]
> > I didn't observe this behavior or maybe didn't pay much attention to it. Just note that the PID values were not optimized yet
> 
> Do you think the bouncing of the arm would cause stow to not execute? This seems like a separate problem that should be addressed to make the code more robust.

I really don't see the issue that you are describing, arm operations seem as expected to me:
[image](https://youtu.be/cPOOEwzRsZo)

**Comment by mogumbo** [comment]
>> Do you think the bouncing of the arm would cause stow to not execute? This seems like a separate problem that should be addressed to make the code more robust.
> 
> I really don't see the issue that you are describing, arm operations seem as expected to me:

Very strange. I see significant bouncing after the first unstow. But it looks like the bouncing is the reason stow cannot execute:

[ERROR] [1606865129.075800743, 1916524833.490000010]: 
Invalid Trajectory: start point deviates from current robot state more than 0.01                                                                              
joint 'j_hand_yaw': expected: 1.27333, current: 1.25636

I think we can wait and see if PID tuning fixes this.

**Review by mogumbo — APPROVED** [review]

**Comment by Samahu** [comment]
> > > Do you think the bouncing of the arm would cause stow to not execute? This seems like a separate problem that should be addressed to make the code more robust.
> > 
> > 
> > I really don't see the issue that you are describing, arm operations seem as expected to me:
> 
> Very strange. I see significant bouncing after the first unstow. But it looks like the bouncing is the reason stow cannot execute:
> 
> [ERROR] [1606865129.075800743, 1916524833.490000010]:
> Invalid Trajectory: start point deviates from current robot state more than 0.01
> joint 'j_hand_yaw': expected: 1.27333, current: 1.25636
> 
> I think we can wait and see if PID tuning fixes this.

I was able to generate the problem you were talking about:

[image](https://www.youtube.com/watch?v=0b53OjrviJ4)

All I had to do is update ow_europa to the latest. This is why I said the physics optimizations that were introduced recently are good for now on the latest ow_europa PR.

**Comment by Samahu** [comment]
resetting the following values to their original values fixes the bouncing issue and stow works
```xml
      <max_step_size>0.0025</max_step_size>
      <real_time_update_rate>400</real_time_update_rate>
```

**Comment by mogumbo** [comment]
Good find. That's too bad. I hope @damianacatanoso can tune things so we can keep the 0.005 step size.

**Review by damianacatanoso — APPROVED** [review]
Tested stow, guarded_move, deliver_sample, unstow with timestep = 0.0025s and update_rate = 400. Everything works well, nice work! Hopefully I'll be able to tune things again to have a higher timestep.

**Comment by Samahu** [comment]
> Tested stow, guarded_move, deliver_sample, unstow with timestep = 0.0025s and update_rate = 400. Everything works well, nice work! Hopefully I'll be able to tune things again to have a higher timestep.

Yes, no doubt! but I think we shouldn't attempt this until PID values have been re-tuned.
